### PR TITLE
Add a possibility to redirect dependency installation messages.

### PR DESCRIPTION
**Pull request by yakutovicha**

### CODE REVIEW

File: aiidalab/app.py [hunk 1]
```diff
@@ -8,14 +8,14 @@
 import sys
 import tarfile
 import tempfile
+import subprocess
```
Comment (csadorf) [inline aiidalab/app.py:11]: Looks like this is not enforced yet, but imports should be alphabetically.

File: aiidalab/app.py [hunk 2]
```diff
@@ -65,6 +65,7 @@ class _AiidaLabApp:
     logo: str = field(default_factory=str)
     categories: List[str] = field(default_factory=list)
     releases: dict = field(default_factory=dict)
+    stdout = sys.stdout
```
Comment (csadorf) [inline aiidalab/app.py:68]: Why are you adding this as a class attribute?

File: aiidalab/app.py [hunk 3]
```diff
@@ -172,11 +173,19 @@ def _install_dependencies(self, python_bin):
...
+                bufsize=1,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
             )
+            while True:
+                output = process.stdout.readline()
+                if output == "" and process.poll() is not None:
+                    break
+                if output:
+                    self.stdout.write(output)
```
Comment (csadorf) [inline aiidalab/app.py:188]: Is this a recipe from SO or so? If yes, could you please add a comment with the link?

File: aiidalab/app.py [hunk 4]
```diff
@@ -494,6 +503,9 @@ def _has_git_repo(self):
         except NotGitRepository:
             return False
 
+    def set_stdout(self, stdout):
+        self._app.stdout = stdout
+
```
Comment (csadorf) [inline aiidalab/app.py:508]: I don't think this should be a class attribute or property, but be passed directly to the function that uses it. Could you comment on the design choice, please?

**Review by csadorf — CHANGES_REQUESTED** [review]

File: aiidalab/app.py (hunk 1 above)
Comment (yakutovicha) [reply aiidalab/app.py]: done

File: aiidalab/app.py (hunk 3 above)
Comment (yakutovicha) [reply aiidalab/app.py:190]: added comment.

File: aiidalab/app.py (hunk 4 above)
Comment (yakutovicha) [reply aiidalab/app.py:510]: My motivation here was to limit the amount of code needed to achieve the goal of redirecting the output. The chain of function calls that would be needed to transfer the `stdout` arguments is quite long, IMO:

`AppManagerWidget._install_version` -> `AiidaLabApp.install_app()` -> `_AiidaLabApp.install()` -> `_AiidaLabApp._install_dependencies()` -> `_pip_install()`

Instead, I thought it would make sense to redirect all the kind-of-log messages to a single output stream once.

That was my motivation. @csadorf what are the advantages of passing stdout as a function argument?

File: aiidalab/app.py (hunk 2 above)
Comment (yakutovicha) [reply aiidalab/app.py:68]: I answered this question in the comment below.

File: aiidalab/app.py (hunk 4 above)
Comment (csadorf) [reply aiidalab/app.py:510]: Ok, so it should certainly not be a class attribute. 

Making it a class variable implies that it is part of the object's state, whereas in reality where you want to direct the output of a specific function calls is ephemeral and only relevant for the specific invocation of that function. You would also expect that the stream cannot be changed during the function execution, right? This design pattern also introduces potential race conditions, since you would need to lock the instance to ensure that the output stream is not changed while you are executing whatever function is requiring output.

In short, avoiding the chain of arguments is not a sufficient justification to introduce this pattern in my opinion.

So my recommendation would be to simply "bite the bullet" and pass the stream argument down the chain. Adding a simple `out=None` argument to the relevant functions is really not that complex. Should there be a second case where we need to redirect some stream, we could consider to refactor this for the `HasTrait` class, but at the moment I would vote against that.

File: aiidalab/app.py (hunk 2 above)
Comment (csadorf) [reply aiidalab/app.py:68]: responded

File: aiidalab/app.py (hunk 4 above)
Comment (yakutovicha) [reply aiidalab/app.py:510]: Ok, I agree with your reasoning.

**Review by csadorf — APPROVED** [review]
LGTM!
